### PR TITLE
Andrew/test coverage

### PR DESCRIPTION
**Pull request by ayjayt**
This PR implements testing for the entire public API.

It also fixes all the bugs found by that testing.

Bugs Fixed:

* Allows users to pass file paths for plotlyjs/mathjax in url and path form
* Explicitly handle mathjax=True
* Better handle instances where users pass `Path` instead of `str`: (casting and `isinstance` flexibility)
* close browser before closing kaleido, prevents hanging in certain instances
* use named arguments in Kaleido constructor instead of generic kwargs (smell fix)
* create temporary file during `.open()` given that it requires `close()` (prevents leak in instances where user doesn't `.open()`, real corner case)
* properly check for an error log before trying to `append` to it.
* deal with corner case where `asyncio.return_task()` doesn't return at ask

These were all bugs found from doing tests using the public API as its expected to be used.

Also, typing was added, a private attribute was changed to public to aid testing, and tests are added.

### CODE REVIEW

File: src/py/kaleido/_page_generator.py [hunk 1]
```diff
@@ -54,7 +56,14 @@ class PageGenerator:
 """
     """The footer is the HTML that always goes on the bottom. Rarely needs changing."""
 
-    def __init__(self, *, plotly=None, mathjax=None, others=None, force_cdn=False):
+    def __init__(
+        self,
+        *,
+        plotly: None | Path | str | tuple[Path | str, str] = None,
```
Comment (emilykl) [inline src/py/kaleido/_page_generator.py:62]: In what situation could these arguments be tuples? Is that documented anywhere?

Comment (ayjayt) [reply src/py/kaleido/_page_generator.py:62]: I think its probably not documented, but actually, I think plotly.js was getting misinterpreted by the browser as `utf-16`, so i had to set it as `utf-8` manually. So the tuple is `(link, encoding)`.

I think its a hidden feature at the moment, I haven't done a revision of the documentation, I was thinking of making it a `NamedTuple`.

File: src/py/tests/test_page_generator.py [hunk 2]
```diff
@@ -155,37 +189,256 @@ async def test_page_generator():
...
+    assert scripts[0] == DEFAULT_MATHJAX
+    assert scripts[1].endswith("package_data/plotly.min.js")
+    assert scripts[2].endswith("kaleido_scopes.js")
+
+
+async def test_force_cdn():
+    """Test force_cdn=True forces use of CDN plotly even when plotly is available."""
+    # Verify plotly is available first
+    if not find_spec("plotly"):
+        pytest.skip("Plotly not available - cannot test force_cdn override")
```
Comment (emilykl) [inline src/py/tests/test_page_generator.py:222]: Shouldn't this test _only_ run when `plotly` is available for import?

File: src/py/tests/test_page_generator.py [hunk 3]
```diff
@@ -155,37 +189,256 @@ async def test_page_generator():
...
+    scripts, encodings = get_scripts_from_html(without_mathjax)
+
+    assert len(scripts) == 2  # noqa: PLR2004
+    assert scripts[0].endswith("package_data/plotly.min.js")
+    assert scripts[1].endswith("kaleido_scopes.js")
+
+
+# Test user overrides
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(st.data())
```
Comment (emilykl) [inline src/py/tests/test_page_generator.py:246]: How does `@given` and `data.draw()` work? Is it testing all possible values returned from `st_valid_path`, or a random subset?

File: src/py/tests/test_page_generator.py [hunk 4]
```diff
@@ -155,37 +189,256 @@ async def test_page_generator():
...
+
+
+async def test_nonexistent_file_path_raises_error(
+    nonexistent_file_path,
+    nonexistent_file_uri,
+):
+    """Test that nonexistent file paths raise FileNotFoundError."""
+    # Test with regular path
+    with pytest.raises(FileNotFoundError):
+        PageGenerator(plotly=str(nonexistent_file_path))
```
Comment (emilykl) [inline src/py/tests/test_page_generator.py:393]: Should test the case where a `Path` is passed as well, right? For this test and the others.

```suggestion
    # Test with Path object
    with pytest.raises(FileNotFoundError):
        PageGenerator(plotly=nonexistent_file_path)
        
    # Test with path as string
    with pytest.raises(FileNotFoundError):
        PageGenerator(plotly=str(nonexistent_file_path))
```

File: src/py/tests/test_public_api.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+def test_start_stop_sync_server_integration():
+    """Test start_sync_server and stop_sync_server with warning behavior."""
+    # Test starting and stopping with warnings silenced
+    kaleido.start_sync_server(silence_warnings=False)
+
+    # Test starting already started server - should warn
+    with pytest.warns(RuntimeWarning, match="already"):
+        kaleido.start_sync_server(silence_warnings=False)
+
+    kaleido.start_sync_server(silence_warnings=True)
```
Comment (emilykl) [inline src/py/tests/test_public_api.py:184]: We need to explicitly test that no warning is emitted here, right?

Or have you set the Pytest configuration such that a warning will trigger a test failure?

File: src/py/tests/test_page_generator.py (hunk 2 above)
Comment (ayjayt) [reply src/py/tests/test_page_generator.py:222]: Yes, here it says if it can't find plotly, it skips the test. Or "only run when plotly is available"

File: src/py/tests/test_page_generator.py (hunk 3 above)
Comment (ayjayt) [reply src/py/tests/test_page_generator.py:246]: data.draw() is basically what hypothesis does normally, it just lets you do it *during* the test instead of during hypothesis setup. So like all hypothesis testing, it does a random sample of default value n=100.

the `st.data()` + `data.draw()` strategy allows you to do the sampling at runtime during your test function instead of during test setup.. I was having trouble with windows not recognizing `Path(__file__)` as a valid file (according to chat GPT, it can happen), and wanted to use the builtin-in `tmp_path` fixture (pretty robust) which is only available at runtime.

File: src/py/tests/test_public_api.py (hunk 5 above)
Comment (ayjayt) [reply src/py/tests/test_public_api.py:184]: Yes, all warnings are upgraded to errors via CLI options.

File: src/py/tests/test_page_generator.py (hunk 2 above)
Comment (emilykl) [reply src/py/tests/test_page_generator.py:222]: Yeah what I mean is that I don't like the usage of `if (condition): pytest.skip()` because it could lead to essentially silently skipping this test entirely if Plotly is never available when the test is run. 

Can't this `if` check be removed entirely?

File: src/py/tests/test_page_generator.py [hunk 6]
```diff
@@ -13,138 +18,167 @@
...
+    assert file_path.resolve().exists()
+    _valid_file_string = str(file_path.resolve())
 
+    _h_file_str = st.just(_valid_file_string)
+    _h_file_path = st.just(Path(_valid_file_string))
+    _h_file_uri = st.just(Path(_valid_file_string).as_uri())
+
+    _h_url = st.tuples(
+        st.sampled_from(["s", ""]),
+        st.text(
```
Comment (emilykl) [inline src/py/tests/test_page_generator.py:147]: might be misunderstanding here since I'm not familiar with `hypothesis`, but the number of possible values that could be sampled here (character strings from 1-20 characters) is huge. Could that space dwarf the other sampling being done here, and reduce the variety of the test cases? I'm not sure if that makes sense, I'm also not sure how `hypothesis` sampling logic works.

Comment (ayjayt) [reply src/py/tests/test_page_generator.py:147]: I've been coming back to this and no, I don't think the range of values makes it more favored in samplying. I think actually for continuous ranges hypothesis uses at least a statistical representation of the possible values (min, max, etc)

File: src/py/tests/test_page_generator.py (hunk 2 above)
Comment (ayjayt) [reply src/py/tests/test_page_generator.py:222]: Okay, but I did it in a later PR
